### PR TITLE
Add ExceptionDispatchInfo.SetRemoteStackTrace

### LINKED ISSUE #50044: API Proposal: ExceptionDispatchInfo.TrySetStackTrace
## Scenario

One of the primary uses of `BinaryFormatter` is to serialize `Exception` objects. Since we are obsoleting this serializer, we need to investigate [new ways of serializing exception data](https://github.com/dotnet/runtime/issues/43482).

For most `Exception`-derived types, serializers can be written to call the applicable public ctor based on parameter name (for `ArgumentException`), HRESULT, exception text, and inner exception object. However, one big omission is the ability to set the stack trace on a newly-constructed, unthrown Exception instance. Without this, "safe" `Exception` serializers and deserializers cannot reproduce the initial exception object with a high enough fidelity to allow for debugging.

## Proposed API

```cs
namespace System.Runtime.ExceptionServices
{
    public sealed class ExceptionDispatchInfo
    {
        // new proposed API
        public static bool TrySetStackTraceString(Exception source, string stackTrace);

        // existing relevant public API, for reference (no changes proposed)
        public static Exception SetCurrentStackTrace(Exception source);
    }
}
```

## Usage

```cs
// Assume deserializer is attempting to reconstruct an InvalidOperationException instance
// from a serialized JSON payload.

var jsonObject = GetJsonObject();

string? message = jsonObject["message"]?.Value;
Exception? innerException = DeserializeExceptionRecursive(jsonObject["innerException"]); // up to some recursive limit
string? stackTrace = jsonObject["stackTrace"]?.Value;
int? hr = jsonObject["hr"]?.IntValue;

var reconstructed = new InvalidOperationException(message, innerException);
if (stackTrace != null) { ExceptionDispatchInfo.TrySetStackTraceString(reconstructed, stackTrace); }
if (hr != null) { reconstructed.HResult = (int)hr; }

// Now you can return the exception object as-is, as a raw deserializer would do
return reconstructed;

// Or you can throw it, as an RPC proxy would do
ExceptionDispatchInfo.Capture(reconstructed).Throw();
```

## Behavior and discussion

The method requires non-null `Exception` and `string` arguments. If the exception instance has been previously thrown or already has an associated stack trace, the method has no effect and returns _false_. (`ExceptionDispatchInfo.SetCurrentStackTrace` would throw in this scenario.) If the exception instance is an immutable or singleton exception, the method has no effect and returns _false_. (`ExceptionDispatchInfo.SetCurrentStackTrace` would no-op in this scenario.) Otherwise the `Exception.StackTrace` property is set to the provided string value and this method returns _true_.

No validation is performed over the contents or shape of the provided string. It is completely opaque to the runtime.

The `Exception.TargetSite` property will remain _null_ after the stack trace string has been set. The runtime will not attempt to parse the stack trace string and recover the original `MethodBase` object. Additionally, the exception object's Watson bucket data will remain unpopulated. This behavior is to support the ideas that: (a) the stack trace might be provided by an untrustworthy source and thus must not be run through the reflection APIs; and (b) the stack trace might refer to an assembly not available to the current process.

## Alternative designs

We could require capturing the exception within the `EDI` before allowing setting the stack trace string. This may even allow us to overwrite the stack trace string from a previously-thrown exception. Doing this would also promote the idea of "this is meant for the purpose of restoring a stack trace from ancillary data before re-throwing the exception" as a first-class concept, even above vanilla deserialization. (The deserialization-without-rethrowing scenario would still be achieveable; simply check the `ExceptionDispatchInfo.SourceException` property instead of rethrowing it.)

However, the existing `SetCurrentStackTrace` method does not require an `EDI` wrapper, so for symmetry the main proposal above also does not require the existence of an `EDI` wrapper.

We could also consider adding a method `SetStackTrace` directly to the `Exception` object. However, I believe this will encourage misuse of what is supposed to be a specialized API. Putting this new API on the `EDI` class makes it a proper sibling of the existing `SetCurrentStackTrace` API, which further helps symmetry.

\>\> __Marked partner blocking__ (/cc @ReubenBond)

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
Resolves https://github.com/dotnet/runtime/issues/50044.

The approved API (see https://github.com/dotnet/runtime/issues/50044#issuecomment-805161306) returns `bool`, which is clearly a mistake given that we wanted to make it a non-try method just like its sibling _SetCurrentStackTrace_. So for this PR, I made the method return `Exception`, just like its sibling. I believe this follows the spirit of what the API review intended.

It's too difficult for me to test the "immutable agile exception" case, so I didn't. That's going to have to remain within the purview of code review.

Also took this time to do some minor cleanup of existing devdoc and unit tests. Nothing too fancy.

/cc @ReubenBond

### CODE REVIEW

**Comment by GrabYourPitchforks** [comment]
Oh no what did I break? :(

File: src/libraries/System.Private.CoreLib/src/System/Runtime/ExceptionServices/ExceptionDispatchInfo.cs [hunk 1]
```diff
@@ -79,5 +79,36 @@ public static Exception SetCurrentStackTrace(Exception source)
...
+        /// <exception cref="ArgumentNullException">The <paramref name="source"/> or <paramref name="stackTrace"/> argument was null.</exception>
+        /// <exception cref="InvalidOperationException">The <paramref name="source"/> argument was previously thrown or previously had a stack trace stored into it.</exception>
+        /// <returns>The <paramref name="source"/> exception instance.</returns>
+        /// <remarks>
+        /// This method populates the <see cref="Exception.StackTrace"/> property from an arbitrary string value.
+        /// The typical use case is the transmission of <see cref="Exception" /> objects across processes with high fidelity,
+        /// allowing preservation of the exception object's stack trace information. .NET does not attempt to parse the
+        /// provided string value. The caller is responsible for normalizing line endings if required.
+        /// </remarks>
+        [StackTraceHidden]
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/ExceptionServices/ExceptionDispatchInfo.cs:97]:
```suggestion
```
Unnecessary

File: src/coreclr/System.Private.CoreLib/src/System/Exception.CoreCLR.cs [hunk 2]
```diff
@@ -325,12 +325,45 @@ internal DispatchState CaptureDispatchState()
...
+            // Store the current stack trace into the "remote" stack trace, which was originally introduced to support
+            // remoting of exceptions cross app-domain boundaries, and is thus concatenated into Exception.StackTrace
+            // when it's retrieved.
+            var sb = new StringBuilder(256);
+            new StackTrace(fNeedFileInfo: true).ToString(System.Diagnostics.StackTrace.TraceFormat.TrailingNewLine, sb);
+            sb.AppendLine(SR.Exception_EndStackTraceFromPreviousThrow);
+            _remoteStackTraceString = sb.ToString();
+        }
+
+        [StackTraceHidden]
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Exception.CoreCLR.cs:343]:
```suggestion
```
Unnecessary

File: src/coreclr/System.Private.CoreLib/src/System/Exception.CoreCLR.cs [hunk 3]
```diff
@@ -325,12 +325,45 @@ internal DispatchState CaptureDispatchState()
...
+
+            // Store the provided text into the "remote" stack trace, following the same format SetCurrentStackTrace
+            // would have generated.
+            _remoteStackTraceString = stackTrace + Environment.NewLineConst + SR.Exception_EndStackTraceFromPreviousThrow + Environment.NewLineConst;
+        }
+
+        // Returns true if setting the _remoteStackTraceString field is legal, false if not (immutable exception).
+        // A false return value means the caller should early-exit the operation.
+        // Can also throw InvalidOperationException if a stack trace is already set or if object has been thrown.
+        [StackTraceHidden]
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Exception.CoreCLR.cs:359]:
```suggestion
```

File: src/mono/System.Private.CoreLib/src/System/Exception.Mono.cs [hunk 4]
```diff
@@ -120,6 +117,26 @@ internal void SetCurrentStackTrace()
             _remoteStackTraceString = sb.ToString();
         }
 
+        [StackTraceHidden]
```
Comment (jkotas) [inline src/mono/System.Private.CoreLib/src/System/Exception.Mono.cs:120]:
```suggestion
```

File: src/mono/System.Private.CoreLib/src/System/Exception.Mono.cs [hunk 5]
```diff
@@ -120,6 +117,26 @@ internal void SetCurrentStackTrace()
...
+        {
+            // Check to see if the exception already has a stack set in it.
+            ThrowIfStackTraceAlreadyExists();
+
+            // Store the provided text into the "remote" stack trace, following the same format SetCurrentStackTrace
+            // would have generated.
+            _remoteStackTraceString = stackTrace + Environment.NewLineConst + SR.Exception_EndStackTraceFromPreviousThrow + Environment.NewLineConst;
+        }
+
+        [StackTraceHidden]
```
Comment (jkotas) [inline src/mono/System.Private.CoreLib/src/System/Exception.Mono.cs:131]:
```suggestion
```

File: src/coreclr/System.Private.CoreLib/src/System/Exception.CoreCLR.cs [hunk 6]
```diff
@@ -325,12 +325,45 @@ internal DispatchState CaptureDispatchState()
...
+        {
+            if (!CanSetRemoteStackTrace())
+            {
+                return; // early-exit
+            }
+
+            // Store the provided text into the "remote" stack trace, following the same format SetCurrentStackTrace
+            // would have generated.
+            _remoteStackTraceString = stackTrace + Environment.NewLineConst + SR.Exception_EndStackTraceFromPreviousThrow + Environment.NewLineConst;
+        }
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Exception.CoreCLR.cs:354]: Can this be in the shared Exception.cs?

(`ThrowIfStackTraceAlreadyExists` can be renamed to CanSetRemoteStackTrace on the Mono side.)

**Review by stephentoub — APPROVED** [review]
Other than Jan's comments, LGTM.

**Review by jkotas — APPROVED** [review]
Thanks!

**Comment by GrabYourPitchforks** [comment]
Opened https://github.com/dotnet/runtime/issues/50520 to track the crossgen2 test failures.

**Pull request merged**